csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubVisitor.cs

```diff
@@ -0,0 +1,835 @@
+                (symbol.ContainingType is not INamedTypeSymbol containingType || containingType.TypeKind != TypeKind.Interface || symbol.IsStatic))
+            {
+                stubWriter.Write("abstract ");
+            }
+        }
+
+        if (symbol.IsStatic)
+            stubWriter.Write("static ");
+        if (symbol.IsVirtual)
+            stubWriter.Write("virtual ");
+        if (symbol.IsOverride)
+            stubWriter.Write("override ");
+        if (symbol.IsSealed)
+        {
+            if (!(symbol is INamedTypeSymbol type && (type.TypeKind == TypeKind.Enum || type.TypeKind == TypeKind.Delegate || type.TypeKind == TypeKind.Struct)))
+                stubWriter.Write("sealed ");
+        }
+        if (symbol.IsExtern)
+            stubWriter.Write("extern ");
+    }
+
+    public void StubTypedConstant(TypedConstant c)
+    {
+        switch (c.Kind)
+        {
+            case TypedConstantKind.Primitive:
+                if (c.Value is string s)
+                {
+                    stubWriter.Write($"\"{s}\"");
+                }
+                else if (c.Value is char ch)
+                {
+                    stubWriter.Write($"'{ch}'");
+                }
+                else if (c.Value is bool b)
+                {
+                    stubWriter.Write(b ? "true" : "false");
+                }
+                else if (c.Value is int i)
+                {
+                    stubWriter.Write(i);
+                }
+                else if (c.Value is long l)
+                {
+                    stubWriter.Write(l);
+                }
+                else if (c.Value is float f)
+                {
+                    stubWriter.Write(f);
+                }
+                else if (c.Value is double d)
+                {
+                    stubWriter.Write(d);
+                }
+                else
+                {
+                    stubWriter.Write("throw null");
+                }
+                break;
+            case TypedConstantKind.Enum:
+                stubWriter.Write("throw null");
+                break;
+            case TypedConstantKind.Array:
+                stubWriter.Write("new []{");
+                WriteCommaSep(c.Values, StubTypedConstant);
+                stubWriter.Write("}");
+                break;
+            default:
+                stubWriter.Write($"/* TODO: {c.Kind} */ throw null");
+                break;
+        }
+    }
+
+    private static readonly HashSet<string> attributeAllowList = new() {
+        "System.FlagsAttribute"
+    };
+
+    private void StubAttribute(AttributeData a, string prefix)
+    {
+        if (a.AttributeClass is not INamedTypeSymbol @class)
+            return;
+
+        var qualifiedName = @class.GetQualifiedName();
+        if (!attributeAllowList.Contains(qualifiedName))
+            return;
+
+        stubWriter.Write($"[{prefix}{qualifiedName.AsSpan(0, @class.GetQualifiedName().Length - 9)}");
```

Comment:
Does `GetQualifiedName` always postfix an attribute with `Attribute` (or does `-9` mean something else)?

Reply:
It is because attribute classes usually (always?) end with `Attribute`, but I guess we may as well handle the case where it doesn't.

Reply:
I think that is just a convention. It is such a strong convention that there is actually compiler support that allows
```csharp
public class MyTestAttribute : Attribute { }

[MyTest]
public class Test1 { }

[MyTestAttribute]
public class Test2 { }
```
However, if you were to write
```csharp
public class MyTest : Attribute { }

[MyTest]
public class Test1 { }

[MyTestAttribute]
public class Test2 { }
```
then you get a compile error only on the attribute usage on `Test2`.